spec/Candid.md

```diff
@@ -882,201 +882,150 @@ service { <name> : <functype>; <methtype>;* } <: service { <name> : <functype'>;
-Only records coerce at record type. Missing fields of option or reserved type turn into `null`.
+In the following rule:
+
+ * The `<nat1>*` field names are those present in both the actual and expected type. The values are coerced.
+ * the `<nat2>*` field names those only in the actual type. The values are ignored.
+ * the `<nat3>*` field names are those only in the expected type, which therefore must be of optional or reserved type. The `null` value is used for these.
 
-In the following rule, the `<nat1>*` field names are those present in both the value and the type, the `<nat2>*` field names only those in the value, and `<nat3>*` are those only in the type.
 ```
-<v1> ~> <v1'> : <t1>
-opt empty <: <t3>
----------------------------------------------------------------------------------------------------------------------------------------
-record { <nat1> = <v1>;* <nat2> = <v2>;* } ~> record { <nat1> = <v1'>;* <nat3> = null;* } : record {  <nat1> = <t1>;* <nat3> = <t3>;* }
+c[record { <nat1> = <t1>;* <nat2> = <t2>;* } <: record { <nat1> = <t1'>;* <nat3> = <t3>;* }](record { <nat1> = <v1>;* <nat2> = <v2>;* })
+    = record { <nat1> = c[<t1> <: <t1'>](<v1>);* <nat3> = null;* }
 ```
 
-
 #### Variants
 
 Only a variant value with an expected tag coerces at variant type.
 
 ```
-<v> ~> <v'> : <t>
-----------------------------------------------------------------------------------
-variant { <nat> = <v> } ~> variant { <nat> = <v'> } : variant { <nat> = <t>; _;* }
+c[variant { <nat> = <t>;* _;* } <: variant { <nat> = <t'>;* _;* }](variant { <natN> = <v> })
+    = variant { <natN> = c[<tN> <: <t'N>](<v>) }
 ```
 
 
 #### References
 
-Function and services references coerce unconditionally (note that coercion is only applied when the input value has a subtype of the expected type):
-
-```
-------------------------------------------------------
-func <text>.<id> ~> func <text>.<id> : func <functype>
-```
-
-```
-------------------------------------------------------
-service <text> ~> service <text> : service <actortype>
-```
+Function and services reference values are untyped, so the coercion function is the identity here:
 
 ```
-------------------------------------------------
-principal <text> ~> principal <text> : principal
+c[func <functype> <: func <functype'>](func <text>.<id>) = func <text>.<id>
+c[service <actortype> <: service <actortype'>](service <text>) = service <text>
+c[principal <: principal](principal <text>) = principal <text>
 ```
 
 #### Tuple types
 
 Whole argument and result sequences are coerced with the same rules as tuple-like records. In particular, extra arguments are ignored, and optional parameters read as as `null` if the argument is missing or fails to coerce:
 
 ```
-record {<v>;*} ~> record {<v'>,*} : record {<t>;*}
---------------------------------------------------
-(<v>,*) ~> (<v'>,*) : (<t>,*)
+c[(<t>,*) <: (<t'>,*)]((<v>,*)) = (<v'>,*)
+    if c[record {<t>;*} ~> record {<t'>,*}](record {<v>;*}) = record {<v'>;*}
```

Comment:
Left-over notation?

Reply:
No; this is a way to say “this function does on tuples what it does on records”.

The altnerative would be to write functions `(<t>,*) -> record {<t>;*} ` and `record {<t>;*}  -> (<t>,*)` to explain the bijection; that seemd to be even less clear

Reply:
I was referring to the `~>`.

Reply:
Ah, right!
```suggestion
    if c[record {<t>;*} <: record {<t'>,*}](record {<v>;*}) = record {<v'>;*}
```